tools/witx/src/ast.rs

```diff
@@ -122,10 +183,76 @@ pub struct UnionVariant {
 #[derive(Debug, Clone)]
 pub struct Module {
     pub name: Id,
-    pub definitions: Vec<ModuleDefinition>,
-    pub entries: HashMap<Id, ModuleEntry>,
+    definitions: Vec<ModuleDefinition>,
+    entries: HashMap<Id, ModuleEntry>,
+}
+
+impl Module {
+    pub(crate) fn new(
+        name: Id,
+        definitions: Vec<ModuleDefinition>,
+        entries: HashMap<Id, ModuleEntry>,
+    ) -> Self {
+        Module {
+            name,
+            definitions,
+            entries,
+        }
+    }
+    pub fn import(&self, name: &Id) -> Option<Rc<ModuleImport>> {
+        self.entries.get(name).and_then(|e| match e {
+            ModuleEntry::Import(d) => Some(d.upgrade().expect("always possible to upgrade entry")),
+            _ => None,
+        })
+    }
+    pub fn imports<'a>(&'a self) -> impl Iterator<Item = Rc<ModuleImport>> + 'a {
+        self.definitions.iter().filter_map(|d| match d {
+            ModuleDefinition::Import(d) => Some(d.clone()),
+            _ => None,
+        })
+    }
+    pub fn func(&self, name: &Id) -> Option<Rc<InterfaceFunc>> {
+        self.entries.get(name).and_then(|e| match e {
+            ModuleEntry::Func(d) => Some(d.upgrade().expect("always possible to upgrade entry")),
+            _ => None,
+        })
+    }
+    pub fn funcs<'a>(&'a self) -> impl Iterator<Item = Rc<InterfaceFunc>> + 'a {
+        self.definitions.iter().filter_map(|d| match d {
+            ModuleDefinition::Func(d) => Some(d.clone()),
+            _ => None,
+        })
+    }
 }
 
+impl PartialEq for Module {
+    fn eq(&self, rhs: &Module) -> bool {
+        if self.definitions.len() != rhs.definitions.len() {
+            return false;
+        }
+        for i in self.imports() {
+            if let Some(i_rhs) = rhs.import(&i.name) {
+                if i != i_rhs {
+                    return false;
+                }
+            } else {
+                return false;
+            }
+        }
+        for f in self.funcs() {
+            if let Some(f_rhs) = rhs.func(&f.name) {
+                if f != f_rhs {
+                    return false;
+                }
+            } else {
+                return false;
+            }
+        }
```

Comment:
Are these two loops the same as doing `self.entries == rhs.entries`? If not, could you add a comment about what's different?

Reply:
that's correct, I just didn't think of that. thank you!

Reply:
It looks like this isn't updated yet.